pkg/kubelet/cm/memorymanager/policy_static.go

```diff
@@ -146,53 +152,101 @@ func (p *staticPolicy) Allocate(s state.State, pod *v1.Pod, container *v1.Contai
+		for containerName, containerReusableResources := range containersReusableResources {
+			s.SetMemoryBlocks(podUID, containerName, containerReusableResources)
+		}
+	}
 
-			// the node does not have enough memory, use the node remaining memory and move to the next node
-			requestedSize -= nodeResourceMemoryState.Free
-			nodeResourceMemoryState.Reserved += nodeResourceMemoryState.Free
-			nodeResourceMemoryState.Free = 0
+	return nil
+}
+
+func (p *staticPolicy) updateMachineState(machineState state.NUMANodeMap, numaAffinity []int, resourceName v1.ResourceName, requestedSize uint64) {
+	for _, nodeID := range numaAffinity {
+		machineState[nodeID].NumberOfAssignments++
+		machineState[nodeID].Cells = numaAffinity
+
+		// we need to continue to update all affinity mask nodes
+		if requestedSize == 0 {
+			continue
+		}
+
+		// update the node memory state
+		nodeResourceMemoryState := machineState[nodeID].MemoryMap[resourceName]
+		if nodeResourceMemoryState.Free <= 0 {
+			continue
 		}
+
+		// the node has enough memory to satisfy the request
+		if nodeResourceMemoryState.Free >= requestedSize {
+			nodeResourceMemoryState.Reserved += requestedSize
+			nodeResourceMemoryState.Free -= requestedSize
+			requestedSize = 0
+			continue
+		}
+
+		// the node does not have enough memory, use the node remaining memory and move to the next node
+		requestedSize -= nodeResourceMemoryState.Free
+		nodeResourceMemoryState.Reserved += nodeResourceMemoryState.Free
+		nodeResourceMemoryState.Free = 0
 	}
+}
 
-	s.SetMachineState(machineState)
-	s.SetMemoryBlocks(string(pod.UID), container.Name, containerBlocks)
+func (p *staticPolicy) getPodReusableMemory(pod *v1.Pod, numaAffinity []int, resourceName v1.ResourceName) uint64 {
+	resourceReusableMemory := uint64(0)
+	for _, containersReusableMemory := range p.initContainersReusableMemory[string(pod.UID)] {
+		for _, containerBlocks := range containersReusableMemory {
+			for _, containerBlock := range containerBlocks {
+				if resourceName != containerBlock.Type {
+					continue
+				}
 
-	return nil
+				if containerBlock.Size == 0 {
+					continue
+				}
+
+				if !isNUMAAffinitiesEqual(numaAffinity, containerBlock.NUMAAffinity) {
+					continue
+				}
+
+				resourceReusableMemory += containerBlock.Size
+			}
+		}
+	}
+
+	return resourceReusableMemory
 }
 
 // RemoveContainer call is idempotent
-func (p *staticPolicy) RemoveContainer(s state.State, podUID string, containerName string) error {
-	klog.InfoS("RemoveContainer", "podUID", podUID, "containerName", containerName)
+func (p *staticPolicy) RemoveContainer(s state.State, podUID string, containerName string) {
 	blocks := s.GetMemoryBlocks(podUID, containerName)
 	if blocks == nil {
-		return nil
+		return
 	}
 
+	klog.InfoS("RemoveContainer", "podUID", podUID, "containerName", containerName)
 	s.Delete(podUID, containerName)
 
+	// delete pod re-usable memory
+	delete(p.initContainersReusableMemory, podUID)
```

Comment:
Hmm. I guess it will work solely because `initContainersReusableMemory ` is only used during allocation, and *all* allocations happen before `AddContainer` (or subsequently RemoveContainer`) are called. It still feels unintuitive though (especially since it will only do anything when it is called on the first cpntainer, and then be a no-op for all other containers).

Reply:
It doesn't seem like this was moved yet?

Reply:
I added it under `updateInitContainersReusableMemory`, but forgot to delete it here.